pkg/cloudprovider/provider/vsphere/vmgroup.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	clusterv1alpha1 "github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+
+	"github.com/vmware/govmomi/vim25/types"
+	"go.uber.org/zap"
+)
+
+func (p *provider) addToVMGroup(ctx context.Context, log *zap.SugaredLogger, session *Session, machine *clusterv1alpha1.Machine, config *Config) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Check if the VM group exists
+	vmGroup, err := findVMGroup(ctx, session, config.Cluster, config.VMGroup)
+	if err != nil {
+		return err
+	}
+
+	// We have to find all VMs in the folder and add them to the VM group. VMGroup only contains VM reference ID which is not enough to
+	// identify the VM by name.
+	machineSetName := machine.Name[:strings.LastIndex(machine.Name, "-")]
+	vmsInFolder, err := session.Finder.VirtualMachineList(ctx, strings.Join([]string{config.Folder, "*"}, "/"))
+	if err != nil {
+		return fmt.Errorf("failed to find VMs in folder: %w", err)
+	}
+
+	var vmRefs []types.ManagedObjectReference
+	for _, vm := range vmsInFolder {
+		// Only add VMs with the same machineSetName to the rule and exclude the machine itself if it is being deleted
+		if strings.HasPrefix(vm.Name(), machineSetName) && !(vm.Name() == machine.Name && machine.DeletionTimestamp != nil) {
+			vmRefs = append(vmRefs, vm.Reference())
+		}
+	}
+
+	var vmRefsToAdd []types.ManagedObjectReference
+	for _, vm := range vmRefs {
+		for _, existingVM := range vmGroup.Vm {
+			if existingVM.Value == vm.Value {
+				log.Debugf("VM %s already in VM group %s", machine.Name, config.VMGroup)
+				continue
+			}
+		}
+		vmRefsToAdd = append(vmRefsToAdd, vm)
```

Comment:
In this case you append very VM to VMRefsToAdd regardless of whether it was already in the group or not

Reply:
Yeah I mistakenly added continue instead of break :)